Introduce hash table switch strategy

fwbrasil opened the pull request:
This is the first change related to https://github.com/oracle/graal/issues/893.

## Problem

The current switch strategies don't work well when the keys are sparse values since the density of the table switch is too low and Graal has to choose less efficient strategies.

## Solution

Introduce a new switch strategy based on cheap hash functions. The `Hasher` evaluates the available functions and tries to find one that guarantees no conflicts for the provided keys. The approach uses imperfect hash functions based on the paper ["Improving Switch Statement Performance with Hashing Optimized at Compile Time"](http://programming.sirrida.de/hashsuper.pdf).

`HashTableSwitchOp` emits the assembly code for the hash table switch. If there's a default target, it also emits the original keys as part of the table and verifies the value to handle possible conflicts with unknown keys.

I've done some [benchmarking](https://github.com/fwbrasil/scala-graal/blob/ed07f1a249b99a3fa89397392bfa9f5c15d56125/src/main/scala/bench/HashSwitchBench.java#L122) and, in some cases, the hash table strategy has similar performance to a regular table switch but with better compaction:

```
BinaryStrategy[avgEffort=6.416389351081533] (the default)

Benchmark                    Mode  Cnt         Score        Error  Units
HashSwitchBench.httpStatus  thrpt   20  26146297.275 ± 345930.644  ops/s

Table switch (density 0.15048543689320387)

Benchmark                    Mode  Cnt         Score        Error  Units
HashSwitchBench.httpStatus  thrpt   20  32873184.212 ± 384754.780  ops/s

Hasher[function=x, effort=1, cardinality=224] (density 0.2767857142857143)

Benchmark                    Mode  Cnt         Score        Error  Units
HashSwitchBench.httpStatus  thrpt   20  32400323.086 ± 218257.904  ops/s
```

Even though the hash table strategy can achieve better compaction than the regular table switch for some cases, the current heuristic used to decide between table switches and other strategies isn't very robust and rarely chooses the hash switch. This issue will be addressed in another PR as described in https://github.com/oracle/graal/issues/893

## Notes

1. Is it ok to implement the new strategy only for one architecture?
2. I don't have much experience writing assembly. This first version doesn't seem very optimized since it uses more registers and uses two moves when `defaultTarget` is not null. I'm also not sure if the padding is ok. Suggestions are appreciated :)

christhalinger commented on compiler/src/org.graalvm.compiler.lir/src/org/graalvm/compiler/lir/gen/LIRGenerator.java:
```diff
@@ -449,29 +455,58 @@ public Variable emitForeignCall(ForeignCallLinkage linkage, LIRFrameState frameS
+        double minDensity = 1 / Math.sqrt(strategy.getAverageEffort() * tableSwitchBonus);
+
+        Optional<Hasher> hasher = Optional.empty();
+        if (supportsHashTableSwitch()) {
```
Having a method `

christhalinger replied:
Having a `supportsHashTableSwitch()` seems odd. Could we instead add a method that returns the hasher and the default implementation just returns `Optional.empty()`?

christhalinger commented:
> Is it ok to implement the new strategy only for one architecture?

Yes, that's ok. Some things are only implemented for AMD64. We just need to find a good way to abstract it (see my comment).

dougxc commented on compiler/src/org.graalvm.compiler.lir/src/org/graalvm/compiler/lir/gen/LIRGenerator.java:
```diff
@@ -88,6 +90,8 @@
 
     public static class Options {
         // @formatter:off
+        @Option(help = "", type = OptionType.Expert)
```
New options must not have an empty `help` value. That said, it would be best to avoid adding this option altogether. Can you not tune the heuristic to choose the hashing strategy?

dougxc commented on compiler/src/org.graalvm.compiler.lir/src/org/graalvm/compiler/lir/hashing/HashFunction.java:
```diff
@@ -0,0 +1,142 @@
+
+abstract class HashFunction {
+
+    public abstract int apply(long value, long s);
```
What does the parameter `s` mean?

dougxc commented on compiler/src/org.graalvm.compiler.lir/src/org/graalvm/compiler/lir/hashing/Hasher.java:
```diff
@@ -0,0 +1,125 @@
+import jdk.vm.ci.meta.JavaConstant;
+import jdk.vm.ci.meta.Value;
+
+public final class Hasher {
```
Please javadoc this new class and its methods.

tkrodriguez commented on compiler/src/org.graalvm.compiler.lir.amd64/src/org/graalvm/compiler/lir/amd64/AMD64ControlFlow.java:
```diff
@@ -314,6 +314,97 @@ public void emitCode(CompilationResultBuilder crb, AMD64MacroAssembler masm) {
+            masm.leaq(scratchReg, new AMD64Address(AMD64.rip, 0));
+            final int afterLea = masm.position();
+
+            // Jump to default target if index doesn't contain the expected value
```
Some comments explaining the table structure would be helpful and maybe merge the defaultTarget != null ifs.  I think you might be able merge the leaq/movslq into one instruction, either by patching the movslq instead or by precomputing the expected offset to the jump table.

tkrodriguez commented on compiler/src/org.graalvm.compiler.lir.amd64/src/org/graalvm/compiler/lir/amd64/AMD64ControlFlow.java:
```diff
@@ -314,6 +314,97 @@ public void emitCode(CompilationResultBuilder crb, AMD64MacroAssembler masm) {
+
+            // Jump to default target if index doesn't contain the expected value
+            if (defaultTarget != null) {
+                masm.movslq(valueScratchReg, new AMD64Address(scratchReg, idxScratchReg, Scale.Times8, 0));
```
you might consider loading 8 bytes here instead of issuing two loads.  If you put key is in the low 32 bits you can still use cmpl and then just perform a shift instead of the second load.  You'd want to align the table at 8 bytes instead which might be a good idea in any case when you have both key and offset in the table.  Having a single load would also allow you to eliminate the move of indexReg to idxScratchReg since you'd only have a single use of it.

tkrodriguez commented on compiler/src/org.graalvm.compiler.lir/src/org/graalvm/compiler/lir/hashing/Hasher.java:
```diff
@@ -0,0 +1,125 @@
+                }
+            }
+            if (candidates.isEmpty()) {
+                return Optional.empty();
```
Any idea how often this happens in practice?

fwbrasil replied:
It happens often since `minDensity` is normally too low and the hasher stops the evaluation based on it. It's necessary to review the heuristic that makes the decision between the density of the table switch x the cost of other strategies. I'll remove the bonus parameter for now and follow up with a separate PR.

tkrodriguez commented:
In your benchmark numbers what kind of switch implementation is being used in the "without hash switch" case?  Presumably it's not a tableswitch.  how does that perform?

plokhotnyuk commented:
I hope it will fix the following performance regression: https://github.com/oracle/graal/issues/412

fwbrasil commented:
@plokhotnyuk I doesn't seem it will. The values are too sparse and the hasher can only find a function at arity 846, which would make the table too large.

```java
public static void main(String[] args) {
        int[] keys = {3355, -1193609971, 3373707, -43264386, 1901043637, -1724546052, 116079, -2102114367, -608539730, -460163995, -1666926107, -1974943731, 1369680106, 226316666, 1699658638,
                        36848094, -1085743021, -1994383672, -1824078800, 3314158, 647263482, 583435343, 1690715422, 1592986856, -1189803120, 594447228, 1021718665, -174080651, -782008927, -690137550,
                        1512257332, -1412739479, 486184128, -1774134745, -101767272, 284519336, -312724053, 545330631, 765915793, 585125142, 1272354024, -475640257};
        JavaConstant[] c = new JavaConstant[keys.length];
        for (int i = 0; i < keys.length; i++)
            c[i] = JavaConstant.forInt(keys[i]);
        Optional<Hasher> forKeys = forKeys(c, 0.00001);
        System.out.println(forKeys); // Optional[Hasher[function=rotateRight(x, p) ^ x, effort=5, cardinality=864]]
    }
```

fwbrasil commented:
@christhalinger @tkrodriguez @dougxc thank you for the reviews! I've just addressed your suggestions.

@tkrodriguez I changed the code to force specific strategies to collect the performance numbers:

```
BinaryStrategy[avgEffort=6.416389351081533] (the default)

Benchmark                    Mode  Cnt         Score        Error  Units
HashSwitchBench.httpStatus  thrpt   20  26146297.275 ± 345930.644  ops/s


Table switch (density 0.15048543689320387)

Benchmark                    Mode  Cnt         Score        Error  Units
HashSwitchBench.httpStatus  thrpt   20  32873184.212 ± 384754.780  ops/s


Hasher[function=x, effort=1, cardinality=224] (density 0.2767857142857143)

Benchmark                    Mode  Cnt         Score         Error  Units
HashSwitchBench.httpStatus  thrpt   20  65606834.026 ± 1701259.405  ops/s
```

I'm puzzled that the hash table switch performs much better than the table switch. I did a bunch of manual testing and it seems to function properly, but I wonder if there's something wrong.

fwbrasil commented:
okay, I just found the bug. The register annotations were wrong. The hash table switch has performance similar to the regular table switch after the fix:

```
Hasher[function=x, effort=1, cardinality=224] (density 0.2767857142857143)

Benchmark                    Mode  Cnt         Score        Error  Units
HashSwitchBench.httpStatus  thrpt   20  32400323.086 ± 218257.904  ops/s
```

fwbrasil commented:
I've introduced tests and addressed the initial review comments. The PR is ready for review.

dougxc commented on compiler/src/org.graalvm.compiler.lir/src/org/graalvm/compiler/lir/hashing/Hasher.java:
```diff
@@ -0,0 +1,161 @@
+import jdk.vm.ci.meta.Value;
+
+/**
+ * This class holds a hash function at a specific cardinality and min value (lowest key).
```
Can you please explain cardinality a bit more in this doc.

dougxc commented on compiler/src/org.graalvm.compiler.jtt/src/org/graalvm/compiler/jtt/optimize/Switch03TestGenerator.java:
```diff
@@ -0,0 +1,112 @@
+
+import jdk.vm.ci.meta.JavaConstant;
+
+class Switch03TestGenerator {
```
It's fine to leave it in. I would suggest making it a static nested class named `Generator` at the bottom of `Switch03.java` to make the connection as clear as possible.

fwbrasil commented:
@dougxc I've just pushed changes to address your comments and added a new test that will fail if the set of hash functions changes and the test is not updated.

tkrodriguez commented:
Why did the performance drop so much after your changes?  Was it executing incorrectly so it seemed faster than it really was?

tkrodriguez commented on compiler/src/org.graalvm.compiler.lir/src/org/graalvm/compiler/lir/gen/LIRGenerator.java:
```diff
@@ -449,29 +451,55 @@ public Variable emitForeignCall(ForeignCallLinkage linkage, LIRFrameState frameS
         double tableSwitchDensity = keyCount / (double) valueRange;
+
         /*
          * This heuristic tries to find a compromise between the effort for the best switch strategy
```
Do we still have a policy problem here?  The existing policy is hard to make sense of but it seems like it's not really making comparable trade offs between table switch and the other forms.  The table forms are only billed for their space requirements but the other forms are billed for their dynamic work and not for their code size.  Binary search for instance has a log scale effort but doubles in size each time which doesn't seem quite fair.

fwbrasil replied:
yes, this heuristic needs to be improved. I'm planning to create a new PR with a better one.

tkrodriguez commented on compiler/src/org.graalvm.compiler.lir/src/org/graalvm/compiler/lir/gen/LIRGenerator.java:
```diff
@@ -449,29 +451,55 @@ public Variable emitForeignCall(ForeignCallLinkage linkage, LIRFrameState frameS
          * gradually with additional effort.
          */
-        if (strategy.getAverageEffort() < 4 || tableSwitchDensity < (1 / Math.sqrt(strategy.getAverageEffort()))) {
+        if (strategy.getAverageEffort() < 4d || (tableSwitchDensity < minDensity && hashTableSwitchDensity < minDensity)) {
```
Shouldn't this be || instead of &&?  If the hashTableSwitchDensity is good but tableSwitchDensity isn't wouldn't we still want to use the hash?

fwbrasil replied:
The condition seems correct: 
- if both tables have density lower than `minDensity`, use the strategy switch
- if one of them is higher than `minDensity` , use one of the table switches

tkrodriguez replied:
Yes of course you're right.  I was just misreading the logic.

tkrodriguez commented on compiler/src/org.graalvm.compiler.lir/src/org/graalvm/compiler/lir/gen/LIRGenerator.java:
```diff
@@ -449,29 +451,55 @@ public Variable emitForeignCall(ForeignCallLinkage linkage, LIRFrameState frameS
+
         int keyCount = keyConstants.length;
         SwitchStrategy strategy = SwitchStrategy.getBestStrategy(keyProbabilities, keyConstants, keyTargets);
+
```
too much vertical whitespace.

tkrodriguez commented on compiler/src/org.graalvm.compiler.lir.amd64/src/org/graalvm/compiler/lir/amd64/AMD64ControlFlow.java:
```diff
@@ -283,9 +283,7 @@ public void emitCode(CompilationResultBuilder crb, AMD64MacroAssembler masm) {
-            if ((masm.position() & 0x3) != 0) {
-                masm.nop(4 - (masm.position() & 0x3));
-            }
+            masm.align(4);
```
I think you might want to align to 8 bytes if defaultTarget != null since each entry is 8 bytes wide.

fwbrasil replied:
That's the alignment for the regular table switch, it hasn't changed. Here's the alignment for the hash table switch: https://github.com/oracle/graal/pull/895/files/7341ab80f998290de7310f393b4191a9f509813c#diff-a8baf9fe5005ee604d024bb2e35c173aR371

dougxc commented:
All my comments have been sufficiently addressed - thanks!

fwbrasil commented:
@dougxc @tkrodriguez @christhalinger thank you for the reviews! I believe all review comments were addressed, please let me know if there's anything pending to merge this change.

fwbrasil commented:
any updates on this? I'd like to create the second PR with a better heuristic to chose the switch strategy

fwbrasil commented:
@dougxc @tkrodriguez since I've planned opening multiple PR for https://github.com/oracle/graal/issues/893 and github doesn't support dependent pull requests, I'm currently blocked. Could you clarify how the merge process work so I can plan accordingly? I imagine it's not as simple as just hitting the merge button here on github.

dougxc commented:
@tkrodriguez can you please integrate this PR.

tkrodriguez commented:
Sorry I hadn't responded earlier.  It has to be run through out internal gate which does a much more extensive set of tests.  I think there's also a question of whether this is a win overall even though it looks good on a microbenchmark.  I'll run our benchmarks and see how it looks as well.  So it requires a bit more attention from me and I've been juggling a few other things.  Sorry for not communicating that.

You can always create a new PR that includes these changes and then merge and rebase it away once it lands.  Git makes that very easy.

tkrodriguez commented on compiler/src/org.graalvm.compiler.lir/src/org/graalvm/compiler/lir/gen/ArithmeticLIRGeneratorTool.java:
```diff
@@ -44,6 +44,8 @@
  */
 public interface ArithmeticLIRGeneratorTool {
 
+    LIRGenerator getLIRGen();
```
Can we remove this override?  It sort of defeats the purpose of the interface to expose this here.  I think replacing ArithmeticLIRGeneratorTool with ArithmeticLIRGenerator in Hasher and HashFunction resolve the problem without requiring this override.

fwbrasil replied:
good point! it does work, I've removed the override

fwbrasil commented:
@tkrodriguez branch rebased

tkrodriguez commented on compiler/src/org.graalvm.compiler.jtt/src/org/graalvm/compiler/jtt/optimize/Switch03.java:
```diff
@@ -0,0 +1,635 @@
+ * Tests optimization of hash table switches.
+ * Code generated by `Switch03.TestGenerator.main`
+ */
+public class Switch03 extends JTTTest {
```
Please use a more descriptive name since this is specifically about the hash switch, like HashSwitchTest.  Also is there any reason this is a JTTest?  These are special variants of regular compiler tests that are exercised with their parameters bounds as constants in addition to a normal compile to test high level transforms.  I don't think you care about that here.  I think this should probably be over in org.graalvm.compiler.core.test.  Or if you wanted it to be a more specific test of the LIR you could make it a LIRTest and check for the existence of a HashTableSwitchOp in the final LIR.  Anyway, renaming is enough for now.  You can make any improvements in a follow on PR.

tkrodriguez commented on compiler/src/org.graalvm.compiler.lir/src/org/graalvm/compiler/lir/gen/LIRGenerator.java:
```diff
@@ -449,29 +451,51 @@ public Variable emitForeignCall(ForeignCallLinkage linkage, LIRFrameState frameS
+                    keys[idx] = keyConstants[i];
+                    targets[idx] = keyTargets[i];
+                }
+                emitHashTableSwitch(hasher.get(), keys, defaultTarget, targets, value);
```
hasher.get is just h isn't it?

tkrodriguez commented on compiler/src/org.graalvm.compiler.lir/src/org/graalvm/compiler/lir/hashing/Hasher.java:
```diff
@@ -0,0 +1,163 @@
+        Set<Integer> seen = new HashSet<>(keys.length);
+        for (JavaConstant key : keys) {
+            int hash = function.apply(key.asLong(), min) & (cardinality - 1);
+            if (seen.contains(hash)) {
```
Can't this be simplified with the else removed?
```suggestion
            if (!seen.add(hash)) {
```

tkrodriguez commented:
The benchmark results look ok.  It doesn't trigger very often from what I can see.  Should we add an option to force it's use?  That could be in a follow on PR if you like.

tkrodriguez commented on compiler/src/org.graalvm.compiler.lir/src/org/graalvm/compiler/lir/hashing/HashFunction.java:
```diff
@@ -0,0 +1,170 @@
+
+    private static Value rotateRight(ArithmeticLIRGenerator gen, Value i, Value distance) {
+        // (i >>> distance) | (i << -distance)
+        return gen.emitOr(gen.emitUShr(i, distance), gen.emitShl(i, gen.emitNegate(distance)));
```
AMD64ArithmeticLIRGenerator provides emitRor which would be a lot more efficient.  Maybe provide a platform independent definition like above and use emitRor for AMD64?

fwbrasil replied:
@tkrodriguez maybe I could add `emitRor` to `ArithmeticLIRGenerator` with the slow implementation as the default and then `AMD64ArithmeticLIRGenerator` overrides it with a better impl?

fwbrasil commented:
@tkrodriguez I've just addressed your last comments.

> The benchmark results look ok. It doesn't trigger very often from what I can see. Should we add an option to force it's use? That could be in a follow on PR if you like.

Yes, I have a branch with a new heuristic that will make the hash table switch more competitive. I'll create the PR once this change is merged. The binary strategy has a large code footprint, so the hash table strategy will be chosen in some cases where the binary one is chosen at the moment.

tkrodriguez commented:
Ok.  I'm going to do some copyright data adjustments in a new commit and push it.

tkrodriguez commented on compiler/src/org.graalvm.compiler.lir/src/org/graalvm/compiler/lir/hashing/HashFunction.java:
```diff
@@ -0,0 +1,165 @@
+            }
+
+            private Value toDWORD(Value v, ArithmeticLIRGenerator t) {
+                if (v.getPlatformKind() != AMD64Kind.DWORD) {
```
This use of AMD64Kind in a shared file should be fixed in the next PR.  I would in general expect the values to come into this function with the right sizing of either int or long with the caller ensuring that it was at least those types already.

dougxc replied:
Why wait for the next PR? Best to fix this here.

fwbrasil replied:
Let me fix it. Should I create a new PR with the fix or are you going to reopen this one?

tkrodriguez replied:
Create a new PR.
